### PR TITLE
[MCP] Add the Create_Record Built in tool

### PR DESCRIPTION
**Pull request by aaronburtle**
## Why make this change?

Closes https://github.com/Azure/data-api-builder/issues/2828

By adding the `Create_Record` tool to the MCP endpoint.

## What is this change?
Add a built in tool, `Create_Record` which uses the `rest` infrastructure to form and validate the query needed to complete a create action. Behavior is at parity with the rest equivalent create operation.

## How was this tested?

Manually tested Insomnia using post.
image

image

image

## Sample Request(s)

```
{
  "jsonrpc": "2.0",
  "id": 2,
  "method": "tools/call",
  "params": {
    "name": "create_record",
    "arguments": {
      "entity": "Broker",
      "data": {
        "ID Number": 3,
				"First Name": "Michael",
				"Last Name": "Jordan"
      }
    }
  }
}
 
```

### CODE REVIEW

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 1]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
             };
         }
 
-        public Task<CallToolResult> ExecuteAsync(
+        public async Task<CallToolResult> ExecuteAsync(
             JsonDocument? arguments,
             IServiceProvider serviceProvider,
             CancellationToken cancellationToken = default)
         {
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs:58]: We need to add a check to see if the tool is exposed or not, and return an error message accordingly.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 2]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                    try
+                    {
+                        requestValidator.ValidateInsertRequestContext(insertRequestContext);
+                    }
+                    catch (DataApiBuilderException ex)
+                    {
+                        return new CallToolResult
+                        {
+                            Content = [new TextContentBlock { Type = "text", Text = $"Error: Request validation failed: {ex.Message}" }]
+                        };
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Add `IsError = true` to return it as error instead of success.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 3]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                else if (result is ObjectResult objectResult)
                 {
-                    Content = [new TextContentBlock { Type = "text", Text = result }]
-                });
+                    return new CallToolResult
+                    {
+                        Content = [new TextContentBlock 
+                        { 
+                            Type = "text", 
+                            Text = $"Record creation completed with status {objectResult.StatusCode}. Result: {JsonSerializer.Serialize(objectResult.Value)}"
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Just to confirm, are we returning the same response of hoe REST returns for create_record?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 4]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                ["error"] = new Dictionary<string, object?>
+                {
+                    ["type"] = errorType,
+                    ["message"] = message
+                }
+            };
+
+            string output = JsonSerializer.Serialize(errorObj);
+
+            logger?.LogWarning("UpdateRecordTool error {ErrorType}: {Message}", errorType, message);
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: This should be create_record tool

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 5]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+
+            return new CallToolResult
+            {
+                Content =
+                [
+                    new TextContentBlock { Type = "text", Text = output }
+                ]
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Same, add `IsError = true`

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 6]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                    insertPayloadRoot,
+                    EntityActionOperation.Insert);
 
-                // TODO: Implement actual create logic using DAB's internal services
-                // For now, return a placeholder response
-                string result = $"Would create record in entity '{entityName}' with data: {dataElement.GetRawText()}";
+                RequestValidator requestValidator = serviceProvider.GetRequiredService<RequestValidator>();
+                
+                // Only validate tables
+                if (dbObject.SourceType is EntitySourceType.Table)
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs:145]: When the type is not table, may be we could error out with that reason, to give more context for the MCP Client.

**Comment by anushakolan** [comment]
The current PR description only covers the success case. Since the tool should also be validated against edge cases and other standard scenarios, I recommend expanding the testing section to include details on the various test cases executed to confirm the tool’s expected behavior.

**Comment by anushakolan** [comment]
The unit tests are failing, please run dotnet format and that should fix most of the syntax issues.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 7]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
             CancellationToken cancellationToken = default)
         {
+            ILogger<CreateRecordTool>? logger = serviceProvider.GetService<ILogger<CreateRecordTool>>();
             if (arguments == null)
             {
-                return Task.FromResult(new CallToolResult
-                {
-                    Content = [new TextContentBlock { Type = "text", Text = "Error: No arguments provided" }]
-                });
+                return BuildErrorResult("Invalid Arguments", "No arguments provided", logger);
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: We already merged the PR that has the global functions. Will you change it in this PR or will you refactor it in a follow up?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 8]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
                 {
-                    return Task.FromResult(new CallToolResult
-                    {
-                        Content = [new TextContentBlock { Type = "text", Text = "Error: Missing required arguments 'entity' or 'data'" }]
-                    });
+                    return BuildErrorResult("Invalid Arguments", "Missing required arguments 'entity' or 'data'", logger);
                 }
 
                 string entityName = entityElement.GetString() ?? string.Empty;
+                if (string.IsNullOrEmpty(entityName))
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: I think it woud be better to check if it is whitespace as well
```suggestion
                if (string.IsNullOrWhitespace(entityName))
```

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 9]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                {
+                    return BuildErrorResult("Invalid Arguments", "Entity name cannot be empty", logger);
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: In the other PRs we have the `ErrorType` without any spaces. Please change it for this one an all the others
```suggestion
                    return BuildErrorResult("InvalidArguments", "Entity name cannot be empty", logger);
```

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 10]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                {
+                    return BuildErrorResult("Invalid Configuration", "Runtime configuration not available", logger);
+                }
+
+                string dataSourceName;
+                try
+                {
+                    dataSourceName = runtimeConfig.GetDataSourceNameFromEntityName(entityName);
+                }
+                catch (DataApiBuilderException)
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Is a `DataApiBuilderException` the only possible exception? I think it would be better to also at a `catch (Exception)`

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 11]
```diff
@@ -37,51 +53,218 @@ public Tool GetToolMetadata()
...
+                
+
+                // Create an HTTP context for authorization
+                IHttpContextAccessor httpContextAccessor = serviceProvider.GetRequiredService<IHttpContextAccessor>();
+                HttpContext httpContext = httpContextAccessor.HttpContext ?? new DefaultHttpContext();
+                IAuthorizationResolver authorizationResolver = serviceProvider.GetRequiredService<IAuthorizationResolver>();
+
+                if (httpContext is null || !authorizationResolver.IsValidRoleContext(httpContext))
+                {
+                    return BuildErrorResult("PermissionDenied", "Permission denied: unable to resolve a valid role context for update operation.", logger);
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: nit: 
```suggestion
                    return BuildErrorResult("PermissionDenied", "Permission denied: Unable to resolve a valid role context for update operation.", logger);
```

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 12]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                // Only validate tables
+                if (dbObject.SourceType is EntitySourceType.Table)
                 {
-                    Content = [new TextContentBlock { Type = "text", Text = result }]
-                });
+                    try
+                    {
+                        requestValidator.ValidateInsertRequestContext(insertRequestContext);
+                    }
+                    catch (DataApiBuilderException ex)
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Same as comment above. Is it only possible for the function to return a `DataApiBuilderException`?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 13]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                    {
+                        return new CallToolResult
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: I think we should use the `BuildErrorResult` to output this error.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 14]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                        }]
+                    };
+                }
+                else if (result is ObjectResult objectResult)
+                {
+                    return new CallToolResult
+                    {
+                        Content = [new TextContentBlock 
+                        { 
+                            Type = "text",
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Is this expected to be an error or can it also be something else?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 15]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                    return new CallToolResult
+                    {
+                        Content = [new TextContentBlock { Type = "text", Text = $"Successfully created record in entity '{entityName}'" }]
+                    };
+                }
             }
             catch (Exception ex)
             {
-                return Task.FromResult(new CallToolResult
+                return new CallToolResult
```
Comment (RubenCerna2079) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: This should also be called by `BuildErrorResult`

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 16]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                    };
+                }
+                else if (result is ObjectResult objectResult)
+                {
+                    return new CallToolResult
+                    {
+                        Content = [new TextContentBlock 
+                        { 
+                            Type = "text", 
+                            Text = $"Record creation completed with status {objectResult.StatusCode}. Result: {JsonSerializer.Serialize(objectResult.Value)}"
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: If this is a success case, why is the message different here. Why is it not: "Successfully created record in entity"

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 14 above)
Comment (Aniruddh25) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Along the same lines, if its an error, IsError = true should be set.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 17]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                    return true;
+                }
+            }
+
+            error = "You do not have permission to create records for this entity.";
+            return false;
+        }
+
+        private static CallToolResult BuildErrorResult(
+        string errorType,
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: nit: should be tabs before each argument declaration - similar to other function definitions.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 18]
```diff
@@ -37,51 +52,235 @@ public Tool GetToolMetadata()
...
+                return new CallToolResult
                 {
                     Content = [new TextContentBlock { Type = "text", Text = $"Error: {ex.Message}" }]
-                });
+                };
             }
         }
+
+        private static bool TryResolveAuthorizedRole(
+        HttpContext httpContext,
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: nit: should have tabs before the argument names

**Review by Aniruddh25 — APPROVED** [review]
Need to reuse the refactored helper methods (perhaps in a followup - please create a task for that), but can be merged as long as comments are addressed.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 7 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Ill use the one we already have in main.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 14 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: This may or may not be an error, so I now have it looking at the status code to determine if IsError should be true or not, and a more generic message.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 16 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: It might not be a success case.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 19]
```diff
@@ -37,51 +51,200 @@ public Tool GetToolMetadata()
...
+                        Content = [new TextContentBlock 
+                        { 
+                            Type = "text", 
+                            Text = $"Successfully created record in entity '{entityName}'. Result: {JsonSerializer.Serialize(createdResult.Value)}"
+                        }]
+                    };
+                }
+                else if (result is ObjectResult objectResult)
+                {
+                    // Check if this is an error status code (400+ range)
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Are all the ObjectResult cases where the record has been created? Is there additional information that can be appended to the successful creation message?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 20]
```diff
@@ -37,51 +51,200 @@ public Tool GetToolMetadata()
...
+                    return new CallToolResult
+                    {
+                        Content = [new TextContentBlock
+                        {
+                            Type = "text",
+                            Text = $"Successfully created record in entity '{entityName}'. Result: {JsonSerializer.Serialize(objectResult.Value)}"
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: If its not an ObjectResult, is it still a successful case?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 19 above)
Comment (Aniruddh25) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Don't we need to check for different error status codes before implying it was always a success?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 3 above)
Comment (anushakolan) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: > Just to confirm, are we returning the same response of hoe REST returns for create_record?

Waiting on this, rest looks good to me.

**Comment by anushakolan** [comment]
The unit tests are failing because of formatting issues, fix the formatting and re-run the tests. Approving the changes to unblock the PR, but waiting on
1. confirming if the tool response matches the REST create record response.
2. All tests are successfully passing.

**Review by anushakolan — APPROVED** [review]

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 3 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: yes we take the same results as the REST return and simply wrap it in the MCP format for the calltoolresult.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 19 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: added the chceks for these edge cases.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 20 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: added a null case here that is an error, otherwise it could be a complete operation with unexpected results, in which case we indicate that in our response,

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 21]
```diff
@@ -172,31 +171,49 @@ public async Task<CallToolResult> ExecuteAsync(
                 {
                     return new CallToolResult
                     {
-                        Content = [new TextContentBlock 
-                        { 
-                            Type = "text", 
+                        Content = [new TextContentBlock
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Should use the McpResponseBuilder.BuildSuccessResult.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 22]
```diff
@@ -172,31 +171,49 @@ public async Task<CallToolResult> ExecuteAsync(
...
                 }
                 else if (result is ObjectResult objectResult)
                 {
-                    // Check if this is an error status code (400+ range)
+                    bool isError = objectResult.StatusCode.HasValue && objectResult.StatusCode.Value >= 400 && objectResult.StatusCode.Value != 403;
+                    string text = isError
+                        ? $"Failed to create record in entity '{entityName}'. Error: {JsonSerializer.Serialize(objectResult.Value)}"
+                        : $"Successfully created record in entity '{entityName}'. Result: {JsonSerializer.Serialize(objectResult.Value)}. Unable to perform " +
+                        $"read-back of inserted records";
                     return new CallToolResult
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Should use BuildErrorResult in case of `isError` otherwise use BuildSuccessResult

**Review by Aniruddh25 — APPROVED** [review]

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 23]
```diff
@@ -169,30 +169,39 @@ public async Task<CallToolResult> ExecuteAsync(
...
+                    }
+                    else
+                    {
+                        return Utils.McpResponseBuilder.BuildSuccessResult(
+                            new Dictionary<string, object?>
+                            {
+                                ["entity"] = entityName,
+                                ["operation"] = "create",
+                                ["result"] = objectResult.Value,
+                                ["note"] = "Unable to perform read-back of inserted records"
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: is `note` a standard expected response property?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs [hunk 24]
```diff
@@ -205,14 +214,16 @@ public async Task<CallToolResult> ExecuteAsync(
...
+                        return Utils.McpResponseBuilder.BuildSuccessResult(
+                            new Dictionary<string, object?>
                             {
-                                Type = "text",
-                                Text = $"Create operation completed for entity '{entityName}' with unexpected result type: {result.GetType().Name}"
-                            }]
-                        };
+                                ["entity"] = entityName,
+                                ["operation"] = "create",
+                                ["result_type"] = result.GetType().Name,
```
Comment (Aniruddh25) [inline src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Is `result_type` a standard expected response?

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 23 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: Actually looks like we are using "message" here in delete records, so I changed now to align with that.

File: src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs (hunk 24 above)
Comment (aaronburtle) [reply src/Azure.DataApiBuilder.Mcp/BuiltInTools/CreateRecordTool.cs]: I removed this and appended it to the "message"

**Review by Aniruddh25 — APPROVED** [review]
